src/io/flutter/server/vmService/VMServiceManager.java

```diff
@@ -252,21 +270,25 @@ private void onVmServiceReceived(String streamId, Event event) {
   }
 
   private void maybeAddServiceExtension(String name) {
-    synchronized (pendingServiceExtensions) {
+    synchronized (serviceExtensions) {
```

Comment:
Just curious as to why we need to synchronize on serviceExtensions here vs pendingServiceExtensions?

Reply:
to avoid future multi-threading bugs. Generally, the fact that we use synchronized a lot in this class isn't great. For this sort of code to be robust it is best to have some simple invariants. Specifically, the one I want is that we never acquire more than one "lock" at a time. Previously you could acquire a couple different locks at once. The simplest way to get a deadlock is if thread 1 locks object A and then tries to lock object B while thread 2 locks object B and then tries to lock object A.  If you only write code that locks one object at a time then that sort of case can never happen.